crates/bevy_ui/src/render/mod.rs

```diff
@@ -382,10 +392,10 @@ pub fn prepare_uinodes(
 ) {
     ui_meta.vertices.clear();
 
-    // sort by increasing z for correct transparency
+    // sort by ui stack index, starting from the deepest node
     extracted_uinodes
         .uinodes
-        .sort_by(|a, b| FloatOrd(a.transform.w_axis[2]).cmp(&FloatOrd(b.transform.w_axis[2])));
+        .sort_by_key(|node| node.stack_index);
```

Comment:
Its a bit unfortunate that we already have an absolute order in UiStack, which we're then re-establishing here in uinodes. The sort is redundant. That being said, if we want a "distributed" extract (ex: multiple systems with node-type-specific extraction logic), then the only real alternative is extracting nodes as entities (instead of pushing them onto an unsorted ExtractedUiNodes list), then using an extracted UiStack value to query/iterate over them in `prepare_uinodes`. The extra cost of "per-entity extraction" might outweigh the time saved on sorting.

Worth measuring, but not something we need to block on here, especially given that this is following the pre-existing pattern.